flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -526,18 +510,16 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
         }
         Long materializationID = materializationIdByCheckpointId.remove(checkpointId);
         if (materializationID != null) {
-            Set<Long> checkpoints = pendingMaterializationConfirmations.get(materializationID);
-            checkpoints.remove(checkpointId);
-            if (checkpoints.isEmpty()) {
```

Comment:
The purpose of maintaining the `Set` of checkpoints was to not notfiy of abortion if any of the checkpoints is still not confirmed.
For example, there is
1. single materialization
1. two pending checkpoints using it
1. 1st get's aborted (and notified)
1. later, 2nd get's confirmed (and notified)

In (3), we'd notify of abortion and of completion in (4) - of the same materialization.
The Set would prevent this.

I don't see how the updated version addresses this. Or am I missing something?

Reply:
I think the newly added `lastCandidateAbortedMaterializationId` could prevent this and would not notify abortion in (3) and then notify completion in (4).

Reply:
How about this case:
1. materialization `M1` completes
1. checkpoint 1 using `M1` is started and aborted - `lastCandidateAbortedMaterializationId` is set to `1`
1. materialization `M2` completes
1. checkpoint `2` using `M2` is started and aborted - `lastCandidateAbortedMaterializationId` is set to `2`, nested backend notified of **abortion** of `2`
1. checkpoint `3` using `M2` is started and completed - `lastConfirmedMaterializationId` is set to `3`, nested backend notified of **completion** of `2`

Reply:
Actually, step 4 would not happen.
When we abort `chk-2`, the nested backend would be notified as abortion of `materialization-id-1`, and then the `lastCandidateAbortedMaterializationId` would be set as `2`.

Reply:
I see, I think you are right. 
I was confused by using `materializationID` in the 1st check and `lastCandidateAbortedMaterializationId` in the actual call.